### PR TITLE
CLDR : Replace legacy number and price formatting

### PR DESCRIPTION
**Pull request by LittleBigDev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Replace the old CLDR implementation with the new one
| Type?         | new feature
| Category?     | CO
| BC breaks?    | no
| Deprecations? | yes
| Fixed ticket? | [FF-165](http://forge.prestashop.com/browse/FF-165)
| How to test?  | All pages where you see numbers and prices (FO and BO)

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8935)

### CODE REVIEW

**Comment by eternoendless** [comment]
Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[classes/Tools.php, line 676 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pXlNqBCxQ6FNZjvHX:-L9pXlNqBCxQ6FNZjvHY:b-8wz8j0) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/classes/Tools.php#L676)):*
> ```PHP
>      * Returns a well formatted number
>      *
>      * @deprecated Please use Locale::formatNumber() instead
> ```

You should add "since 1.7.4.0" in the deprecation notice, and then trigger a deprecation error in the method:
`@trigger_error('XXX() is deprecated since version YYY. Use XXX instead.', E_USER_DEPRECATED);`

---

*[config/services/admin/services_prod.yml, line 72 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pYVfa64NZI5n_akMj:-L9pYVfa64NZI5n_akMk:b-dge1l6) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/config/services/admin/services_prod.yml#L72)):*
> ```YAML
>             - '@prestashop.core.localization.cldr.middleware.topLayer'
> 
>     prestashop.core.localization.cldr.middleware.topLayer: '@prestashop.core.localization.cldr.middleware.localeCache'
> ```

It should be "datalayer" instead of "middleware"

---

*[config/services/front/services_prod.yml, line 45 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pbr_BBahDUi1dBJvh:-L9pbr_BBahDUi1dBJvi:bn420if) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/config/services/front/services_prod.yml#L45)):*
> ```YAML
>             - '@prestashop.core.localization.cldr.middleware.topLayer'
> 
>     prestashop.core.localization.cldr.middleware.topLayer: '@prestashop.core.localization.cldr.middleware.localeCache'
> ```

Again, middleware -> datalayer

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935)*

**Comment by LittleBigDev** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[classes/Tools.php, line 676 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pXlNqBCxQ6FNZjvHX:-L9shapp57oTX3k2uMwZ:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/classes/Tools.php#L676)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

You should add "since 1.7.4.0" in the deprecation notice, and then trigger a deprecation error in the method:
`@trigger_error('XXX() is deprecated since version YYY. Use XXX instead.', E_USER_DEPRECATED);`

</blockquote></details>

Done.

---

*[config/services/admin/services_prod.yml, line 72 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pYVfa64NZI5n_akMj:-L9si7wrBSjrb1pnRQNZ:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/config/services/admin/services_prod.yml#L72)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

It should be "datalayer" instead of "middleware"
</blockquote></details>

Done.

---

*[config/services/front/services_prod.yml, line 45 at r1](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9pbr_BBahDUi1dBJvh:-L9si8U20YxqB5Z9hob3:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/4cd0793975511177bc34c7d11cec3f6c7e08f013/config/services/front/services_prod.yml#L45)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Again, middleware -> datalayer
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935)*

**Comment by LittleBigDev** [comment]
@eternoendless wdyt ?

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 3 of 3 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-:-L9tEoth3RXzSPAWHu2J:bnfp4nl)*

**Review by eternoendless — DISMISSED** [review]

**Comment by marionf** [comment]
@LittleBigDev 

I wanted to test the stats page in BO because there is a lot of numbers, but I have an error on develop branch
I can't validate it without checking stats page :(

capture du 2018-04-12 16-04-04

**Comment by LittleBigDev** [comment]
Fixed. Before testing again, please proceed to a new install to get the database fixes !

@eternoendless  can you please review again (easy and ultra fast review !) ?

**Comment by eternoendless** [comment]
:lgtm:

---

Reviewed 2 of 2 files at r4, 1 of 1 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-:-L9uMrOrDC0rHggj7nF_:bnfp4nl)*

**Comment by LittleBigDev** [comment]
Rebased to fix conflicts with `develop`.

File: classes/Tools.php [hunk 1]
```diff
@@ -641,42 +645,96 @@ public static function getCldr(Context $context = null, $language_code = null)
...
+        if (is_int($currency)) {
+            $currency = Currency::getCurrencyInstance($currency);
         }
-        if ($currency === null) {
-            $currency = $context->currency;
-        } elseif (is_int($currency)) {
-            $currency = Currency::getCurrencyInstance((int)$currency);
+
+        $container = $context->controller->getContainer();
+        if (null === $container) {
```
Comment (mickaelandrieu) [inline classes/Tools.php]: what will happens if SymfonyContainer::getInstance returns null?

This function should be moved to Adapter/Tools class imo

File: classes/Tools.php [hunk 2]
```diff
@@ -641,42 +645,96 @@ public static function getCldr(Context $context = null, $language_code = null)
...
+            $container = SymfonyContainer::getInstance();
         }
 
-        $cldr = self::getCldr($context);
+        /** @var LocaleRepository $localeRepository */
+        $localeRepository = $container->get(self::SERVICE_LOCALE_REPOSITORY);
+        $locale           = $localeRepository->getLocale((string)$context->language->locale);
```
Comment (mickaelandrieu) [inline classes/Tools.php]: How about inject an instance of "LocaleRepository" in Context class directly? I'm pretty sure you are already in a Controller at this moment, and you dont need to inject container here.

File: classes/Tools.php (hunk 1 above)
Comment (LittleBigDev) [reply classes/Tools.php]: I made sure global `$kernel` is set. So `SymfonyContainer::getInstance()` will not return null.
About moving the function : whatever. It is deprecated anyway and won't be used anymore (ASAP).

File: classes/Tools.php (hunk 2 above)
Comment (LittleBigDev) [reply classes/Tools.php]: Well, sometimes it is possible that the Tools' saved controller is a fake one, having no container (`init()` not called). E.g.: with the "hybrid" AdminProductsController.
I had to find something else

**Comment by LittleBigDev** [comment]
@mickaelandrieu here you are. I can't do much better until Controller's `init()` call is 100% reliable.

**Comment by eternoendless** [comment]
Reviewed 6 of 6 files at r7.
Review status: 9 of 12 files reviewed at latest revision, 2 unresolved discussions.

---

*[classes/Context.php, line 95 at r8](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9z40tj0I236rfWb9UO:-L9z40tj0I236rfWb9UP:b-mnqwkp) ([raw file](https://github.com/prestashop/prestashop/blob/7d145ef4ea93fd6b5406ab6d1b17975903a2a42b/classes/Context.php#L95)):*
> ```PHP
>      * @var LocaleRepository
>      */
>     public $localeRepository;
> ```

Wouldn't it make more sense to store the current Locale instance instead of the Repository? The class is called Context after all.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935)*

**Comment by LittleBigDev** [comment]
Review status: 9 of 12 files reviewed at latest revision, 3 unresolved discussions.

---

*[classes/Context.php, line 95 at r8](https://beta.reviewable.io/reviews/prestashop/prestashop/8935#-L9z40tj0I236rfWb9UO:-L9z72Us8hRxHAeN7mz7:b-896fix) ([raw file](https://github.com/prestashop/prestashop/blob/7d145ef4ea93fd6b5406ab6d1b17975903a2a42b/classes/Context.php#L95)):*
<details><summary><i>Previously, eternoendless (Pablo Borowicz) wrote…</i></summary><blockquote>

Wouldn't it make more sense to store the current Locale instance instead of the Repository? The class is called Context after all.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/prestashop/prestashop/8935)*

**Comment by LittleBigDev** [comment]
Finally reverted the Locale instance injection. It implies too much changes in legacy unit test classes and we have no time for that.
Sticking to locale repository.

**Comment by LittleBigDev** [comment]
@mickaelandrieu and @eternoendless  WDYT ?

**Review by mickaelandrieu — APPROVED** [review]

**Comment by mickaelandrieu** [comment]
Code reviewed, QA reviewed and tests OK => LGTM

**Comment by mickaelandrieu** [comment]
Bip bip!

Just noticed your pull request introduces some new notices in our PHPUnit test suite => any idea how to fix it?

https://travis-ci.org/PrestaShop/PrestaShop/jobs/370046902#L922

**Comment by LittleBigDev** [comment]
@Quetzacoalt91 has a PR for this :) => https://github.com/PrestaShop/PrestaShop/pull/8978

**Comment by mickaelandrieu** [comment]
@Quetzacoalt91 my man, my hero \o/

**Comment by Quetzacoalt91** [comment]
Wait wait wait, I'd be not so sure about that.
The linked PR fixes notices thrown during the upgrade process, these notices are new to me.

**Comment by LittleBigDev** [comment]
Oh sorry, my bad :open_mouth: 
Looks like I have to open a new PR for these notices

**Comment by LittleBigDev** [comment]
Oh sorry, my bad :open_mouth: 
Looks like I have to open a new PR for these notices

**Pull request merged**